rfcs/2020-09-28-cloudwatch-metrics.md

```diff
@@ -0,0 +1,181 @@
+    period_secs = 300 # period (s) to aggregate metrics over; defaults to top-level `period` setting; top-level interval should be a multiple of this and any other defined periods
+```
+
+We could alternatively model dimensions as another table:
+
+```toml
+[[sources.cloudwatch.metrics.dimensions]]
+  key = "InstanceId"
+  value = "i-05517fbc2e6124dfb"
+```
+
+To support:
+
+* globbing metric names
+* fetching all metrics for a namespace
+* globbing of dimension values
+
+We will refresh and cache the available metrics on an interval
+(`metrics_refresh_interval`) for namespaces appearing in the configuration.
+
+A wildcard dimension value (`"*"`) will pull and publish discrete metrics for
+each dimension value. This is different than omitting the dimension altogether
+which would instead aggregate across the values for the dimension.
+
+Note that dimensions must match the dimensions that the metrics were published
+with. For example, if the metric was published two dimensions,
+availability-zone and load balancer name, then both dimensions must be
+specified to retrieve the metric. CloudWatch won't automatically aggregate
+across the other if only one is specified.
+
+This source will use the default AWS credential chain similar to the
+`aws_cloudwatch_metrics` sink.
+
+The output (shown here in prometheus format) will publish metrics in the form
+of:
+
+```text
+<namespace>_<metric_name>_<metric_stat>{<dimension_key>=<dimension_value>*}
+```
+
+All metrics will be additionally tagged with the region.
+
+For example:
+
+```text
+aws_ec2_cpu_credit_balance_average{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_maximum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_minimum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_sample_count{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 1
+aws_ec2_cpu_credit_balance_sum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+```
+
+I think we will want to provide example configs that collect common metrics for
+various namespaces (like EC2, S3, etc.).
+
+## Rationale
+
+Broadening AWS platform support.
+
+Otherwise, users will need to run `telegraf` or another agent to ingest this
+data.
+
+## Prior Art
+
+* [telegraf](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/cloudwatch)
+* [prometheus exporter](https://github.com/prometheus/cloudwatch_exporter)
+* [filebeat](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-cloudwatch.html)
+
+## Drawbacks
+
+The additional maintenance burden of a new source.
+
+## Alternatives
+
+There is alternate API,
+[`GetMetricStatistics'](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricStatistics.html)
+that we could use, but I think it will result in less overall throughput due to
+API rate limit restrictions:
```

Comment:
> due to API rate limit restrictions

Are there any upsides to using `GetMetricsStatistics` that would benefit low-volume usages of this source?

If a sufficient benefit exists, and we think low-volume usage of the source is a relevant use-case, it might make sense to support both endpoints, but default to the one that provides the least amount of friction (e.g. the one without the low rate limit)?

Reply:
The only upside I can think of to using `GetMetricsStatistics` is to avoid taking up API calls to `GetMetricData` (perhaps useful if they have other services already consuming `GetMetricData`). Of course, the opposite is also true if they already have consumers of `GetMetricStatistics`.

Are there others you can think of? Failing any, I think we should just stick to `GetMetricData` rather than complicating the implementation supporting both.